tabbycat/notifications/migrations/0007_migrate_data.py

```diff
@@ -0,0 +1,48 @@
+# Generated by Django 2.0.8 on 2018-10-23 19:44
+# Extended manually to create BulkNotification objects and Message-IDs
+import email
+import re
+
+import django.contrib.postgres.fields.jsonb
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+def create_bulk_notifications(apps, schema_editor):
+    SentMessageRecord = apps.get_model("notifications", "SentMessageRecord")
+    BulkNotification = apps.get_model("notifications", "BulkNotification")
+
+    current_event = None
+    current_bulk = 0
+
+    for m in SentMessageRecord.objects.all():
+        if m.event is not current_event:
+            current_bulk += 1
+            current_event = m.event
+
+            BulkNotification.objects.create(event=m.event, timestamp=m.timestamp,
+                                            round=m.round, tournament=m.tournament)
+        m.notification_id = current_bulk
+        m.save()
```

Comment:
Hmm, it's probably better not to assume that IDs will start at 1 and be ordered sequentially—in most cases they will be, but there are probably edge cases where you end up having to start at a higher number or whatever. Since `.create()` returns the new object, could just look up the ID of the object you just created?

Also, does this loop assume that all `SentMessageRecord` instances relating to the same event are next to each other in the list? Would we want to use `SentMessageRecord.objects.order_by(...)` to guarantee this? Also, `[itertools.groupby()](https://docs.python.org/3/library/itertools.html#itertools.groupby)` might be useful 🙂

Reply:
I haven't really looked into itertools before. Also yes, it does assume that, but I'll correct it.

Reply:
Grouped by timestamp then event. For ballot receipts, made a further conditional. Didn't use itertools.